src/particles.js

```diff
@@ -107,18 +129,133 @@ class Particles extends Layer {
     );
 
     this.initializeParticles(gl, this._numParticles);
+
+    this.nullTexture = util.createTexture(
+      gl,
+      gl.NEAREST,
+      new Uint8Array([0, 0, 0, 0]),
+      1,
+      1
+    );
+
+    this.nullTile = {
+      getTexture: () => this.nullTexture
+    };
   }
 
   // This is a callback from mapbox for rendering into a texture
-  prerender(gl, matrix) {
-    if (this.windData) this.update(gl, matrix);
+  prerender(gl) {
+    if (this.windData) {
+      const blendingEnabled = gl.isEnabled(gl.BLEND);
+      gl.disable(gl.BLEND);
+      const tiles = this.visibleParticleTiles();
+      tiles.forEach(tile => {
+        const found = this.findAssociatedDataTiles(tile);
+        if (found) {
+          this.update(gl, this._particleTiles[tile], found);
+          this._particleTiles[tile].updated = true;
+        }
+      });
+      if (blendingEnabled) gl.enable(gl.BLEND);
+      this.map.triggerRepaint();
+    }
   }
 
-  update(gl) {
-    const blendingEnabled = gl.isEnabled(gl.BLEND);
-    gl.disable(gl.BLEND);
+  /**
+   * This method computes the ideal data tiles to support our particle tiles
+   */
+  computeLoadableTiles() {
+    const result = {};
+    const add = tile => (result[tile] = tile);
+    this.visibleParticleTiles().forEach(tileID => {
+      let t = tileID;
+      let matrix = new DOMMatrix();
+      while (!t.isRoot()) {
+        if (t.z <= this.windData.maxzoom) break;
+        const [x, y] = t.quadrant();
+        matrix.translateSelf(0.5 * x, 0.5 * y);
+        matrix.scaleSelf(0.5);
+        t = t.parent();
+      }
+
+      matrix.translateSelf(-0.5, -0.5);
+      matrix.scaleSelf(2, 2);
+
+      const tl = matrix.transformPoint(new window.DOMPoint(0, 0));
+      const br = matrix.transformPoint(new window.DOMPoint(1, 1));
+
+      add(t);
+
+      if (tl.x < 0 && tl.y < 0) add(t.neighbor(-1, -1));
+      if (tl.x < 0) add(t.neighbor(-1, 0));
+      if (tl.x < 0 && br.y > 1) add(t.neighbor(-1, 1));
+
+      if (br.x > 1 && tl.y < 0) add(t.neighbor(1, -1));
+      if (br.x > 1) add(t.neighbor(1, 0));
+      if (br.x > 1 && br.y > 1) add(t.neighbor(1, 1));
+
+      if (tl.y < 0) add(t.neighbor(0, -1));
+      if (br.y > 1) add(t.neighbor(0, 1));
```

Comment:
Might be better making the values checked against and those passed to neighbour, constants, would give these if statements more context. Basically I'm not sure in this situation what -1, 0, 1 represent.

Reply:
Their offsets of the tile. So given tile z=3, x=2, y=1:

- `tile.neighbor(0, 0) == tile == 3/2/1`
- `tile.neighbor(1, 0) == 3/3/1`
- `tile.neighbor(0, -1) == 3/2/0`

and so on.

Reply:
A comment to that effect would be good then, if others have to support this, its likely to only be every few months anyone will look at the code and this isn't info I at least will remember or clean from the code.